### DISCUSSION TITLE
Quarkus unit testing for Panache active record pattern

### DISCUSSION THREAD

**Discussion by games1232002** [quarkusio/quarkus#37293] [Q&A]
I was having issues writing unit tests for my service using Mockito 5.3.1 while trying to move away from Quarkus Dev Services. The concern I have is that in my use case, there is no need to spin up 2 docker containers containing empty MSSQL databases as it's resource intensive.

For greater certainty, "unit tests" that I'm referring to are small tests for basic parts of my code. Since there's no network connectivity required during those tests, my aim is to disable dev services after removing all dependencies on the MSSQL database and tables **during unit testing**.

I am successful in mocking 90% of my test classes/methods/lines by simply using Mockito, but eventually I will need to mock Panache calls, such as `SomeEntity.listAll()`, `SomeEntity.findById`, `entity.persist()` where `SomeEntity` is a class with the `@Entity` tag and extends `PanacheEntityBase`, while `entity` is an instance of `SomeEntity`.

I am aware that it's documented on https://quarkus.io/guides/hibernate-orm-panache#using-the-active-record-pattern that we can make use of `PanacheMock`, but, correct me if I'm wrong, PanacheMock require the use of dev services.

Since I am using Mockito >= 3.4.0, I have also attempted to use StaticMock by following this guide: https://www.baeldung.com/mockito-mock-static-methods to try and mock the entity class and methods, such as this:
```
try (MockedStatic<SomeEntity> entityMockStatic = Mockito.mockStatic(SomeEntity.class)) {
        entityMockStatic.when(SomeEntity::listAll).thenReturn(Collections.emptyList());
        assertEquals(Collections.emptyList(), SomeEntity.listAll())
    }
```
But that always result in the throwing of `MissingMethodInvocationException` at the `when` line.

After a lot of searching online, and trying different combinations, including reading the following resources:
- https://github.com/quarkusio/quarkus/discussions/27153
- https://github.com/quarkusio/quarkus/discussions/25707

I am still unable to find a way that I can create unit tests without spinning up 2 empty MSSQL database in containers.

Can someone help?

[a comment by an automated account is not shown]

**Reply by ennishol** [reply]
I'd suggest to separate your code logic from panache code. Create a simple service to hide all panache related code. You can then easily mock it and unit test your code.

**Comment by FroMage** [comment]
> correct me if I'm wrong, PanacheMock require the use of dev services

I don't think that's true. This sounds like the perfect solution in your case.

**Reply by games1232002** [reply]
The problem I have is that when dev services are disabled, I have to remove `@QuarkusTest` from the test class, and while other mocks using Mockito are ok, whenever PanacheMock is involved, the test case will fail with this:
```
java.lang.IllegalStateException: This method is normally automatically overridden in subclasses: did you forget to annotate your entity with @Entity?
	at io.quarkus.hibernate.orm.panache.common.runtime.AbstractJpaOperations.implementationInjectionMissing(AbstractJpaOperations.java:443)
	at io.quarkus.hibernate.orm.panache.PanacheEntityBase.find(PanacheEntityBase.java:167)
	...
```

Not sure if this information may help but the class that is mocked using `PanacheMock.mock(SomeEntity.class)` does have the `@Entity` tag, and that class extends `@PanacheEntityBase`.

**Reply by FroMage** [reply]
> when dev services are disabled, I have to remove @QuarkusTest from the test class

Why do you have to remove it?

Most of Quarkus tests will fail when you test it without `@QuarkusTest`, I think.

**Reply by mhagnumdw** [reply]
@FroMage , does `PanacheMock.mock(Holiday.class);` only work with `@QuarkusTest`? I would like to use it with `@QuarkusComponentTest`, because in my tests I would not need Quarkus to upload the entire context, including datasource etc.

**Reply by RomainWilbert** [reply]
I want to do the same as you (mock panache entity static methods without starting the quarkus app). Did you come up with an elegant solution ?
